### PR TITLE
[SPARK-44457][CONNECT][TESTS] Add `truncatedTo(ChronoUnit.MICROS)` to make `ArrowEncoderSuite` in Java 17 daily test GA task pass

### PR DESCRIPTION
**Pull request by LuciferYang**
### What changes were proposed in this pull request?
Similar to SPARK-42770 | https://github.com/apache/spark/pull/40395, this pr call `truncatedTo(ChronoUnit.MICROS)` on `Instant.now()` and `LocalDateTime.now()` to ensure microsecond accuracy is used in any environment.

### Why are the changes needed?
Make Java 17 daily test GA task run successfully. 

The Java 17 daily test GA task failed now: https://github.com/apache/spark/actions/runs/5570003581/jobs/10173767006

```
[info] - nullable fields *** FAILED *** (169 milliseconds)
[info]   NullableData(null, JANUARY, E1, null, 1.000000000000000000, 2.000000000000000000, null, 4, PT0S, null, 2023-07-16, 2023-07-16, null, 2023-07-16T23:01:54.059339Z, 2023-07-16T23:01:54.059359) did not equal NullableData(null, JANUARY, E1, null, 1.000000000000000000, 2.000000000000000000, null, 4, PT0S, null, 2023-07-16, 2023-07-16, null, 2023-07-16T23:01:54.059339538Z, 2023-07-16T23:01:54.059359638) (ArrowEncoderSuite.scala:194)
[info]   Analysis:
[info]   NullableData(instant: 2023-07-16T23:01:54.059339Z -> 2023-07-16T23:01:54.059339538Z, localDateTime: 2023-07-16T23:01:54.059359 -> 2023-07-16T23:01:54.059359638)
[info]   org.scalatest.exceptions.TestFailedException:
...
[info] - lenient field serialization - timestamp/instant *** FAILED *** (26 milliseconds)
[info]   2023-07-16T23:01:55.112838Z did not equal 2023-07-16T23:01:55.112838568Z (ArrowEncoderSuite.scala:194)
[info]   org.scalatest.exceptions.TestFailedException:
...

```

### Does this PR introduce _any_ user-facing change?
No, just for test

### How was this patch tested?
- Pass GitHub Action
- Git Hub Action test with Java 17 passed: https://github.com/LuciferYang/spark/actions/runs/5647253889/job/15297009685

image

### CODE REVIEW

**Comment by LuciferYang** [comment]
https://github.com/apache/spark/actions/runs/5570003581/jobs/10173767006

Java 17 daily test failed

**Comment by LuciferYang** [comment]
```
[info] - nullable fields *** FAILED *** (169 milliseconds)
[info]   NullableData(null, JANUARY, E1, null, 1.000000000000000000, 2.000000000000000000, null, 4, PT0S, null, 2023-07-16, 2023-07-16, null, 2023-07-16T23:01:54.059339Z, 2023-07-16T23:01:54.059359) did not equal NullableData(null, JANUARY, E1, null, 1.000000000000000000, 2.000000000000000000, null, 4, PT0S, null, 2023-07-16, 2023-07-16, null, 2023-07-16T23:01:54.059339538Z, 2023-07-16T23:01:54.059359638) (ArrowEncoderSuite.scala:194)
[info]   Analysis:
[info]   NullableData(instant: 2023-07-16T23:01:54.059339Z -> 2023-07-16T23:01:54.059339538Z, localDateTime: 2023-07-16T23:01:54.059359 -> 2023-07-16T23:01:54.059359638)
[info]   org.scalatest.exceptions.TestFailedException:
[info]   at org.scalatest.Assertions.newAssertionFailedException(Assertions.scala:472)
[info]   at org.scalatest.Assertions.newAssertionFailedException$(Assertions.scala:471)
[info]   at org.scalatest.Assertions$.newAssertionFailedException(Assertions.scala:1231)
[info]   at org.scalatest.Assertions$AssertionsHelper.macroAssert(Assertions.scala:1295)
[info]   at org.apache.spark.sql.connect.client.arrow.ArrowEncoderSuite.$anonfun$compareIterators$1(ArrowEncoderSuite.scala:194)
[info]   at scala.collection.Iterator.foreach(Iterator.scala:943)
[info]   at scala.collection.Iterator.foreach$(Iterator.scala:943)
[info]   at scala.collection.AbstractIterator.foreach(Iterator.scala:1431)
[info]   at org.apache.spark.sql.connect.client.arrow.ArrowEncoderSuite.compareIterators(ArrowEncoderSuite.scala:191)
[info]   at org.apache.spark.sql.connect.client.arrow.ArrowEncoderSuite.roundTripAndCheck(ArrowEncoderSuite.scala:155)
[info]   at org.apache.spark.sql.connect.client.arrow.ArrowEncoderSuite.roundTripAndCheckIdentical(ArrowEncoderSuite.scala:174)
[info]   at org.apache.spark.sql.connect.client.arrow.ArrowEncoderSuite.$anonfun$new$21(ArrowEncoderSuite.scala:359)
[info]   at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
[info]   at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
[info]   at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
[info]   at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
[info]   at org.scalatest.Transformer.apply(Transformer.scala:22)
[info]   at org.scalatest.Transformer.apply(Transformer.scala:20)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike$$anon$1.apply(AnyFunSuiteLike.scala:226)
[info]   at org.scalatest.TestSuite.withFixture(TestSuite.scala:196)
[info]   at org.scalatest.TestSuite.withFixture$(TestSuite.scala:195)
[info]   at org.scalatest.funsuite.AnyFunSuite.withFixture(AnyFunSuite.scala:1564)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike.invokeWithFixture$1(AnyFunSuiteLike.scala:224)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike.$anonfun$runTest$1(AnyFunSuiteLike.scala:236)
[info]   at org.scalatest.SuperEngine.runTestImpl(Engine.scala:306)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike.runTest(AnyFunSuiteLike.scala:236)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike.runTest$(AnyFunSuiteLike.scala:218)
[info]   at org.scalatest.funsuite.AnyFunSuite.runTest(AnyFunSuite.scala:1564)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike.$anonfun$runTests$1(AnyFunSuiteLike.scala:269)
[info]   at org.scalatest.SuperEngine.$anonfun$runTestsInBranch$1(Engine.scala:413)
[info]   at scala.collection.immutable.List.foreach(List.scala:431)
[info]   at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:401)
[info]   at org.scalatest.SuperEngine.runTestsInBranch(Engine.scala:396)
[info]   at org.scalatest.SuperEngine.runTestsImpl(Engine.scala:475)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike.runTests(AnyFunSuiteLike.scala:269)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike.runTests$(AnyFunSuiteLike.scala:268)
[info]   at org.scalatest.funsuite.AnyFunSuite.runTests(AnyFunSuite.scala:1564)
[info]   at org.scalatest.Suite.run(Suite.scala:1114)
[info]   at org.scalatest.Suite.run$(Suite.scala:1096)
[info]   at org.scalatest.funsuite.AnyFunSuite.org$scalatest$funsuite$AnyFunSuiteLike$$super$run(AnyFunSuite.scala:1564)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike.$anonfun$run$1(AnyFunSuiteLike.scala:273)
[info]   at org.scalatest.SuperEngine.runImpl(Engine.scala:535)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike.run(AnyFunSuiteLike.scala:273)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike.run$(AnyFunSuiteLike.scala:272)
[info]   at org.apache.spark.sql.connect.client.arrow.ArrowEncoderSuite.org$scalatest$BeforeAndAfterAll$$super$run(ArrowEncoderSuite.scala:49)
[info]   at org.scalatest.BeforeAndAfterAll.liftedTree1$1(BeforeAndAfterAll.scala:213)
[info]   at org.scalatest.BeforeAndAfterAll.run(BeforeAndAfterAll.scala:210)
[info]   at org.scalatest.BeforeAndAfterAll.run$(BeforeAndAfterAll.scala:208)
[info]   at org.apache.spark.sql.connect.client.arrow.ArrowEncoderSuite.run(ArrowEncoderSuite.scala:49)
[info]   at org.scalatest.tools.Framework.org$scalatest$tools$Framework$$runSuite(Framework.scala:321)
[info]   at org.scalatest.tools.Framework$ScalaTestTask.execute(Framework.scala:517)
[info]   at sbt.ForkMain$Run.lambda$runTest$1(ForkMain.java:414)
[info]   at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
[info]   at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
[info]   at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
[info]   at java.base/java.lang.Thread.run(Thread.java:833)
[info] - lenient field serialization - timestamp/instant *** FAILED *** (26 milliseconds)
[info]   2023-07-16T23:01:55.112838Z did not equal 2023-07-16T23:01:55.112838568Z (ArrowEncoderSuite.scala:194)
[info]   org.scalatest.exceptions.TestFailedException:
[info]   at org.scalatest.Assertions.newAssertionFailedException(Assertions.scala:472)
[info]   at org.scalatest.Assertions.newAssertionFailedException$(Assertions.scala:471)
[info]   at org.scalatest.Assertions$.newAssertionFailedException(Assertions.scala:1231)
[info]   at org.scalatest.Assertions$AssertionsHelper.macroAssert(Assertions.scala:1295)
[info]   at org.apache.spark.sql.connect.client.arrow.ArrowEncoderSuite.$anonfun$compareIterators$1(ArrowEncoderSuite.scala:194)
[info]   at scala.collection.Iterator.foreach(Iterator.scala:943)
[info]   at scala.collection.Iterator.foreach$(Iterator.scala:943)
[info]   at scala.collection.AbstractIterator.foreach(Iterator.scala:1431)
[info]   at org.apache.spark.sql.connect.client.arrow.ArrowEncoderSuite.compareIterators(ArrowEncoderSuite.scala:191)
[info]   at org.apache.spark.sql.connect.client.arrow.ArrowEncoderSuite.roundTripAndCheck(ArrowEncoderSuite.scala:155)
[info]   at org.apache.spark.sql.connect.client.arrow.ArrowEncoderSuite.$anonfun$new$81(ArrowEncoderSuite.scala:594)
[info]   at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
[info]   at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
[info]   at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
[info]   at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
[info]   at org.scalatest.Transformer.apply(Transformer.scala:22)
[info]   at org.scalatest.Transformer.apply(Transformer.scala:20)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike$$anon$1.apply(AnyFunSuiteLike.scala:226)
[info]   at org.scalatest.TestSuite.withFixture(TestSuite.scala:196)
[info]   at org.scalatest.TestSuite.withFixture$(TestSuite.scala:195)
[info]   at org.scalatest.funsuite.AnyFunSuite.withFixture(AnyFunSuite.scala:1564)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike.invokeWithFixture$1(AnyFunSuiteLike.scala:224)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike.$anonfun$runTest$1(AnyFunSuiteLike.scala:236)
[info]   at org.scalatest.SuperEngine.runTestImpl(Engine.scala:306)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike.runTest(AnyFunSuiteLike.scala:236)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike.runTest$(AnyFunSuiteLike.scala:218)
[info]   at org.scalatest.funsuite.AnyFunSuite.runTest(AnyFunSuite.scala:1564)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike.$anonfun$runTests$1(AnyFunSuiteLike.scala:269)
[info]   at org.scalatest.SuperEngine.$anonfun$runTestsInBranch$1(Engine.scala:413)
[info]   at scala.collection.immutable.List.foreach(List.scala:431)
[info]   at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:401)
[info]   at org.scalatest.SuperEngine.runTestsInBranch(Engine.scala:396)
[info]   at org.scalatest.SuperEngine.runTestsImpl(Engine.scala:475)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike.runTests(AnyFunSuiteLike.scala:269)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike.runTests$(AnyFunSuiteLike.scala:268)
[info]   at org.scalatest.funsuite.AnyFunSuite.runTests(AnyFunSuite.scala:1564)
[info]   at org.scalatest.Suite.run(Suite.scala:1114)
[info]   at org.scalatest.Suite.run$(Suite.scala:1096)
[info]   at org.scalatest.funsuite.AnyFunSuite.org$scalatest$funsuite$AnyFunSuiteLike$$super$run(AnyFunSuite.scala:1564)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike.$anonfun$run$1(AnyFunSuiteLike.scala:273)
[info]   at org.scalatest.SuperEngine.runImpl(Engine.scala:535)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike.run(AnyFunSuiteLike.scala:273)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike.run$(AnyFunSuiteLike.scala:272)
[info]   at org.apache.spark.sql.connect.client.arrow.ArrowEncoderSuite.org$scalatest$BeforeAndAfterAll$$super$run(ArrowEncoderSuite.scala:49)
[info]   at org.scalatest.BeforeAndAfterAll.liftedTree1$1(BeforeAndAfterAll.scala:213)
[info]   at org.scalatest.BeforeAndAfterAll.run(BeforeAndAfterAll.scala:210)
[info]   at org.scalatest.BeforeAndAfterAll.run$(BeforeAndAfterAll.scala:208)
[info]   at org.apache.spark.sql.connect.client.arrow.ArrowEncoderSuite.run(ArrowEncoderSuite.scala:49)
[info]   at org.scalatest.tools.Framework.org$scalatest$tools$Framework$$runSuite(Framework.scala:321)
[info]   at org.scalatest.tools.Framework$ScalaTestTask.execute(Framework.scala:517)
[info]   at sbt.ForkMain$Run.lambda$runTest$1(ForkMain.java:414)
[info]   at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
[info]   at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
[info]   at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
[info]   at java.base/java.lang.Thread.run(Thread.java:833)

```

File: .github/workflows/build_and_test.yml [hunk 1]
```diff
@@ -25,7 +25,7 @@ on:
       java:
         required: false
         type: string
-        default: 8
+        default: 17
```
Comment (LuciferYang) [inline .github/workflows/build_and_test.yml:28]: will revert after test

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/arrow/ArrowEncoderSuite.scala [hunk 2]
```diff
@@ -353,8 +355,20 @@ class ArrowEncoderSuite extends ConnectFunSuite with BeforeAndAfterAll {
 
   test("nullable fields") {
     val encoder = ScalaReflection.encoderFor[NullableData]
-    val instant = java.time.Instant.now()
-    val now = java.time.LocalDateTime.now()
+    // SPARK-44457: Similar to SPARK-42770, calling `truncatedTo(ChronoUnit.MICROS)`
+    // for `java.time.Instant` and `java.time.LocalDateTime` when testing with Java 17 on Linux.
+    val needTruncate =
```
Comment (LuciferYang) [inline connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/arrow/ArrowEncoderSuite.scala:360]: We can extract helper functions in a follow up and make `SQLImplicitsTestSuite` reuse the helper functions as well

**Comment by LuciferYang** [comment]
friendly ping @MaxGekk @dongjoon-hyun

**Comment by srowen** [comment]
Can you explain this change? why is the logic different for a specific Java version?

**Comment by LuciferYang** [comment]
@srowen This is not the first case, the first case we can see https://github.com/apache/spark/pull/40395

```
Run `LocalDateTime.now()` and `Instant.now()` with Java 8 & 11 always get microseconds on both Linux and MacOS, but there are some differences when using Java 17, it will get accurate nanoseconds on Linux, but still get the microseconds on MacOS. 

On Linux(CentOs)

```
jshell> java.time.LocalDateTime.now()
$1 ==> 2023-03-13T18:09:12.498162194

jshell> java.time.Instant.now()
$2 ==> 2023-03-13T10:09:16.013993186Z
```

On MacOS

```
jshell> java.time.LocalDateTime.now()
$1 ==> 2023-03-13T17:13:47.485897

jshell> java.time.Instant.now()
$2 ==> 2023-03-13T09:15:12.031850Z
```
 
At present, Spark always converts them to microseconds ...
```

**Comment by srowen** [comment]
Whether it starts with or without microseconds, shouldn't the round-trip encode/decode work? Maybe I misunderstand.
Can we always truncate instead of special-casing?

**Comment by srowen** [comment]
I see, arrow does not encode the full precision? then always truncate here? it is at least simpler.

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/arrow/ArrowEncoderSuite.scala [hunk 3]
```diff
@@ -353,8 +355,20 @@ class ArrowEncoderSuite extends ConnectFunSuite with BeforeAndAfterAll {
...
+      SystemUtils.isJavaVersionAtLeast(JavaVersion.JAVA_17) && SystemUtils.IS_OS_LINUX
+    val instant = if (needTruncate) {
+      java.time.Instant.now().truncatedTo(ChronoUnit.MICROS)
+    } else {
+      java.time.Instant.now()
+    }
+    val now = if (needTruncate) {
```
Comment (LuciferYang) [inline connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/arrow/ArrowEncoderSuite.scala:367]: @srowen For this branch, @dongjoon-hyun  suggested maintaining the behavior for non special cases in the previous pr

https://github.com/apache/spark/pull/40395#discussion_r1134941041

**Comment by LuciferYang** [comment]
> I see, arrow does not encode the full precision? then always truncate here? it is at least simpler.

Let me clarify my thoughts again:

1. This is not related to arrow, the same result applies to using `ExpressionEncoder`（https://github.com/apache/spark/pull/40395）

2. In the current implementation of Spark, we use `DateTimeUtils` to convert `Instant` or `LocalDateTime` to microseconds:

https://github.com/apache/spark/blob/4644344f443f2f6ab26a72b44b9219fb6c82d26e/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala#L542-L568


3. Therefore, in special scenarios, the result of `encoded+decoded` data loses precision after microseconds. 

For example:

Normal scenario：input is `2023-03-13T18:09:12.498162`, the result after `encoded+decoded` is `2023-03-13T18:09:12.498162`, input equals output.

Special scenario（Java17&Linux）: input is `2023-03-13T18:09:12.498162194`, the result after `encoded+decoded` is `2023-03-13T18:09:12.498162`, input not equal to output.

Calling `truncatedTo(ChronoUnit.MICROS)` on input in special scenarios is only for the convenience of comparing results

**Comment by srowen** [comment]
OK, up to your judgment. I'm just wondering if this test is simpler if it never has nanoseconds for any Java version, instead of special-casing.

**Comment by LuciferYang** [comment]
>  I'm just wondering if this test is simpler if it never has nanoseconds for any Java version

Yes

**Comment by LuciferYang** [comment]
> OK, up to your judgment. I'm just wondering if this test is simpler if it never has nanoseconds for any Java version, instead of special-casing.

So change to use a fixed `Instant` or `LocalDateTime` as inputs with only microsecond accuracy?

**Comment by srowen** [comment]
Why not just always truncate? I just didn't see why that's not simpler for this an previous changes. It's valid for earlier Java versions. It makes the result match the desired semantics of the test.

**Comment by LuciferYang** [comment]
In order to only apply new code on Java17 + Linux environment only https://github.com/apache/spark/pull/40395#discussion_r1134941041

But I also think `always truncate` is more simpler.

**Comment by srowen** [comment]
I don't quite understand that logic. We would rather not have different code paths for different environments - although here the difference is trivial.

**Comment by LuciferYang** [comment]
[48b3a05](https://github.com/apache/spark/pull/42039/commits/48b3a053644fcbcb35362b9d081f430327844a84) change to `always truncate`, let's test Java 17 using GA once more

**Comment by LuciferYang** [comment]
rebase

**Review by srowen — APPROVED** [review]

**Comment by srowen** [comment]
Merged to master

**Comment by LuciferYang** [comment]
Thanks @srowen ~

**Pull request closed without merging**